backend/tests/api/test_booking.py

```
@@ -101,17 +131,86 @@ async def test_bookings_all(
+    features = BookingFeature(booked=10, product_feature="hello.world")
+    booking = Booking(job_id=1, features=[features], lead_host="host1", user_name="user1")
+    resp = await backend_client.put("/api/v1/booking/book", json=booking.dict())
+
+    assert resp.status_code == status.HTTP_200_OK
+    edit_counts_mock.assert_awaited_once()
+    map_bookings_mock.assert_awaited_once()
+
+
+@mark.asyncio
+@mock.patch("lm_backend.api.booking.edit_counts")
+@mock.patch("lm_backend.api.booking.map_bookings")
+@database.transaction(force_rollback=True)
+async def test_booking_create_negative_booked_error(
+    edit_counts_mock, map_bookings_mock, backend_client, some_config_rows, some_licenses, insert_objects
+):
+    await insert_objects(some_licenses, table_schemas.license_table)
+    await insert_objects(some_config_rows, table_schemas.config_table)
+    features = BookingFeature(booked=-1, product_feature="hello.world")
+    booking = Booking(job_id=1, features=[features], lead_host="host1", user_name="user1")
+    resp = await backend_client.put("/api/v1/booking/book", json=booking.dict())
+
+    assert resp.status_code == status.HTTP_400_BAD_REQUEST
+    edit_counts_mock.assert_not_awaited()
+    map_bookings_mock.assert_not_awaited()
+
+
+@mark.asyncio
+@database.transaction(force_rollback=True)
+async def test_booking_create_booked_greater_than_total(
+    backend_client, some_config_rows, some_licenses, insert_objects
+):
+    await insert_objects(some_licenses, table_schemas.license_table)
+    await insert_objects(some_config_rows, table_schemas.config_table)
+    features = BookingFeature(booked=1000, product_feature="hello.world")
+    booking = Booking(job_id=1, features=[features], lead_host="host1", user_name="user1")
+    resp = await backend_client.put("/api/v1/booking/book", json=booking.dict())
+
+    assert resp.status_code == status.HTTP_400_BAD_REQUEST
+    assert "<= total" in resp.json()["detail"]
```

```suggestion
@mark.asyncio
@database.transaction(force_rollback=True)
async def test_booking_create_booked_greater_than_total(
        backend_client, some_config_rows, some_licenses, insert_objects,
):
    """This test proves that the correct response (400) is returned when a booking
    request exceeds the total available by asserting that the response detail contains
    the string "<= total" and asserting that the functions map_booking() and edit_counts()
    are not awaited.
    """
    await insert_objects(some_licenses, table_schemas.license_table)
    await insert_objects(some_config_rows, table_schemas.config_table)
    features = BookingFeature(booked=1000, product_feature="hello.world")
    booking = Booking(job_id=1, features=[features], lead_host="host1", user_name="user1")
    resp = await backend_client.put("/api/v1/booking/book", json=booking.dict())

    assert resp.status_code == status.HTTP_400_BAD_REQUEST
    assert "<= total" in resp.json()["detail"]
```

---

This one is better to just call the functions, instead of using mock, because the 400 will raise from then